federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// Update ingress info when service updated, and persist to federation apiserver
+func (cc *clusterClientCache) processServiceUpdate(cachedService *cachedService, service *v1.Service, clusterName string) bool {
+	glog.V(4).Infof("Processing service update for %s/%s, cluster %s", service.Namespace, service.Name, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	var needUpdate bool
+	cachedStatus, ok := cachedService.serviceStatusMap[clusterName]
+	cachedFedServiceStatus := cachedService.lastState.Status.LoadBalancer
+	if ok {
+		if reflect.DeepEqual(cachedStatus, service.Status.LoadBalancer) {
+			glog.V(4).Infof("same ingress info observed for service %s/%s: %+v ", service.Namespace, service.Name, cachedStatus.Ingress)
+		} else{
+			glog.V(4).Infof("ingress info was changed for service %s/%s: %+v ", service.Namespace, service.Name, cachedStatus.Ingress)
+			needUpdate = true
+		}
+	} else {
+		glog.V(4).Infof("cached service status was not found for %s/%s, cluster %s, building one", service.Namespace, service.Name, clusterName)
+		cachedService.serviceStatusMap[clusterName] = service.Status.LoadBalancer
+		needUpdate = true
+
+		// cache is not always reliable(cache will be cleaned when service controller restart)
+		// two cases will run into this branch:
+		// 1. new service loadbalancer info received -> no info in cache, and no in federation service
+		// 2. service controller being restarted -> no info in cache, but it is in federation service
+
+		// check if the lb info is already in federation service
+		CheckLB:
+			for _, fed := range cachedFedServiceStatus.Ingress {
+				for _, new := range service.Status.LoadBalancer.Ingress{
+					if (new.IP == fed.IP && new.Hostname == fed.Hostname) {
+						needUpdate = false
+						break CheckLB
+					}
+				}
+			}
+	}
+
+	if needUpdate {
+		// new status = cached federation status - cached status for the cluster + new status from the cluster
+		for i, fed := range cachedFedServiceStatus.Ingress {
+			for _, new := range cachedStatus.Ingress{
+				// remove if same ingress record found
+				if (new.IP == fed.IP && new.Hostname == fed.Hostname) {
+					cachedFedServiceStatus.Ingress = append(cachedFedServiceStatus.Ingress[:i], cachedFedServiceStatus.Ingress[i+1:]...)
+					glog.V(4).Infof("remove old ingress info %s, %s for service %s/%s", new.Hostname, new.IP, service.Namespace, service.Name)
+				}
+			}
+		}
+		cachedService.lastState.Status.LoadBalancer.Ingress = append(cachedFedServiceStatus.Ingress, service.Status.LoadBalancer.Ingress...)
+		glog.V(4).Infof("add new ingress info %+v for service %s/%s", service.Status.LoadBalancer, service.Namespace, service.Name)
+	} else {
+		glog.V(4).Infof("same ingress info found for %s/%s, cluster %s", service.Namespace, service.Name, clusterName)
+	}
+	return needUpdate
+}
+
+
+func (cc *clusterClientCache) persistFedServiceUpdate(service *v1.Service, fedClient federation_release_1_3.Interface) (error, bool){
```

Comment:
seems we should use internalclient, but cluster controller also uses versioned client. will following.